File: python/src/problems301_400/odd_even_linked_list.py
from abc import ABC, abstractmethod
from typing import override

from src.concepts.singly_list_node import SinglyListNode


class Problem(ABC):
    @abstractmethod
    def odd_even_list(self, head: SinglyListNode | None) -> SinglyListNode | None:
        pass


class Default(Problem):
    @override
    def odd_even_list(self, head: SinglyListNode | None) -> SinglyListNode | None:
        if head is None:
            return None

        odd: SinglyListNode = head
        even: SinglyListNode | None = head.next
        even_head: SinglyListNode = even

        while even and even.next:
            odd.next = even.next
            odd = odd.next
            even.next = odd.next
            even = even.next

        odd.next = even_head
        return head


SOLUTIONS: list[Problem] = [Default()]

File: typescript/src/problems301_400/odd-even-linked-list.ts
import SinglyListNode from '../concepts/singly-list-node';

class Problem {
  oddEvenList(head: SinglyListNode | undefined): SinglyListNode | undefined {
    throw new Error(`Not implemented (${head}).`);
  }
}

class Default extends Problem {
  oddEvenList(head: SinglyListNode | undefined): SinglyListNode | undefined {
    if (!head) {
      return undefined;
    }

    let odd: SinglyListNode = head;
    let even: SinglyListNode | undefined = head.next;
    const evenHead: SinglyListNode = even;

    while (even && even.next) {
      odd.next = even.next;
      odd = odd.next;
      even.next = odd.next;
      even = even.next;
    }

    odd.next = evenHead;
    return head;
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;
